### ISSUE TITLE
ESP32S3 Support

### ISSUE THREAD

**Issue by darnold64** [tkem/CarreraDigitalControlUnit#15]
Thomas,

I have now moved on from the ESP8266 to the ESP32 (which is ultimately were I plan to run things).  However, I encountered a significant problem.  When I ran the code which works fine on the ESP8266 it caused the MCU to constantly recycle/reset itself.  Not wholly unlike the issue I had with the ESP8266.  However, after a lot of troubleshooting I tracked it down to the attachInterrupt(digitalPinToInterrupt(_pin), &irq, CHANGE) statement that appears in the CarreraDigitalControlUnit::start() routine.  I have tried everything I can think of to resolve the issue.  I thought it might be the need to ensure the code was placed in the correct memory location through using the IRAM_ATTR parameter for the interrupt related functions/routines (I recognize the ESP32 can be a bit temperamental when assigning/using interrupts).  Nothing I tried resolved the issue.

In the end, the only way I could get the code to run was to declare the CarreraDigitalControlUnit::irq() routine public and then pull-out the attachInterrupt(digitalPinToInterrupt(_pin), &irq, CHANGE) instruction and place it in the setup() routine in the main body of my sketch (I was actually using your LapCouter example for troubleshooting).  Just before calling the cu.start() I issue the following instruction: attachInterrupt(digitalPinToInterrupt(cuPin), cu.irq, CHANGE);

For the life of me I can't figure out why having the attachInterrupt command in the library file (ie., CarreraDigitalControlUnit.cpp) causes the ESP32 to constantly reset.

**Comment by tkem** [comment]
Frankly, I have no idea. The only thing I can see is that in

https://github.com/tkem/CarreraDigitalControlUnit/blob/fb3e02d51b1ff9b29b7566217c077f23087ea936/src/CarreraDigitalControlUnit.cpp#L22

we should also check for ARDUINO_ARCH_ESP32, since ESP32 also needs IRAM_ATTR. But why the location of the `attachInterrupt()` call (not the interrupt routine itself) would make a difference just beats me.

**Comment by darnold64** [comment]
I had tried that to no avail.  I also tried just directly including IRAM_ATTR in all the interrupt handler routines but still no luck.  The behaviour is clearly a sign that memory is getting stepped on somehow when the attachInterrupt is called within the library.  It must be something to do with how the ESP32 handles objects and where/how it places those in memory.  

Anyway it is very bizarre behaviour.

**Comment by tkem** [comment]
@darnold64: Any updates on this? I just ordered an ESP32 board to see if I can reproduce this, since it sounds so strange ;-)

**Comment by darnold64** [comment]
Sorry, nothing new to report.  My workaround is to modify the library to include a macro that removes/skips the attachInterrupt instruction within the start method based on whether it is an ESP32 or not.  In my main sketch I have a similar macro.  However, this is more of a “patch” rather than a fix.  I did a bit more digging and tried one or two additional things—ie., changing cores but nothing worked.  So I just dropped it for now.  Maybe you will have better luck.
